File: simulation/fault_tolerant_nav.py
"""
Phase 512: Fault-Tolerant Navigation
센서 퓨전 기반 내결함성 항법, IMU/GPS/비전 삼중 중복.
"""

import numpy as np
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict, Optional, Tuple


class NavSensor(Enum):
    GPS = "gps"
    IMU = "imu"
    VISION = "vision"
    BAROMETER = "barometer"
    MAGNETOMETER = "magnetometer"
    LIDAR = "lidar"


class NavHealth(Enum):
    NOMINAL = "nominal"
    DEGRADED = "degraded"
    FAILED = "failed"


@dataclass
class SensorState:
    sensor: NavSensor
    health: NavHealth
    position_est: np.ndarray
    velocity_est: np.ndarray
    covariance: np.ndarray
    confidence: float
    timestamp: float


@dataclass
class NavSolution:
    position: np.ndarray
    velocity: np.ndarray
    heading_deg: float
    confidence: float
    sensors_used: List[NavSensor]
    integrity: float


class ExtendedKalmanFilter:
    """EKF for sensor fusion navigation."""

    def __init__(self, dim: int = 6, seed: int = 42):
        self.rng = np.random.default_rng(seed)
        self.dim = dim
        self.x = np.zeros(dim)  # [px, py, pz, vx, vy, vz]
        self.P = np.eye(dim) * 10.0
        self.Q = np.eye(dim) * 0.01  # process noise
        self.F = np.eye(dim)  # state transition
        self.F[:3, 3:6] = np.eye(3) * 0.1  # dt

    def predict(self, dt: float = 0.1):
        self.F[:3, 3:6] = np.eye(3) * dt
        self.x = self.F @ self.x
        self.P = self.F @ self.P @ self.F.T + self.Q * dt

    def update(self, z: np.ndarray, H: np.ndarray, R: np.ndarray):
        y = z - H @ self.x
        S = H @ self.P @ H.T + R
        try:
            K = self.P @ H.T @ np.linalg.inv(S)
        except np.linalg.LinAlgError:
            return
        self.x = self.x + K @ y
        self.P = (np.eye(self.dim) - K @ H) @ self.P

    def state(self) -> Tuple[np.ndarray, np.ndarray]:
        return self.x[:3].copy(), self.x[3:6].copy()


class SensorVoter:
    """Triple-redundancy sensor voting."""

    def __init__(self):
        self.vote_log: List[Dict] = []

    def vote(self, states: List[SensorState]) -> Tuple[np.ndarray, float]:
        if not states:
            return np.zeros(3), 0.0
        healthy = [s for s in states if s.health != NavHealth.FAILED]
        if not healthy:
            healthy = states  # use all as fallback

        weights = np.array([s.confidence for s in healthy])
        weights /= weights.sum() + 1e-10
        pos = sum(w * s.position_est for w, s in zip(weights, healthy))

        spread = np.std([s.position_est for s in healthy], axis=0)
        integrity = max(0, 1 - np.linalg.norm(spread) / 10)

        self.vote_log.append({"n_sensors": len(healthy), "integrity": round(integrity, 4)})
        return pos, integrity


class FaultTolerantNav:
    """Fault-tolerant navigation system with sensor fusion."""

    def __init__(self, seed: int = 42):
        self.rng = np.random.default_rng(seed)
        self.ekf = ExtendedKalmanFilter(seed=seed)
        self.voter = SensorVoter()
        self.sensor_health: Dict[NavSensor, NavHealth] = {
            s: NavHealth.NOMINAL for s in NavSensor}
        self.solutions: List[NavSolution] = []
        self.true_pos = np.array([0.0, 0.0, 50.0])
        self.true_vel = np.array([2.0, 1.0, 0.0])

    def _simulate_sensor(self, sensor: NavSensor, dt: float) -> SensorState:
        health = self.sensor_health[sensor]
        noise_map = {
            NavSensor.GPS: 2.0, NavSensor.IMU: 0.5, NavSensor.VISION: 1.0,
            NavSensor.BAROMETER: 3.0, NavSensor.MAGNETOMETER: 5.0, NavSensor.LIDAR: 0.3,
        }
        noise = noise_map.get(sensor, 1.0)
        if health == NavHealth.DEGRADED:
            noise *= 5
        elif health == NavHealth.FAILED:
            noise *= 50

        pos_est = self.true_pos + self.rng.standard_normal(3) * noise
        vel_est = self.true_vel + self.rng.standard_normal(3) * noise * 0.3
        cov = np.eye(3) * noise ** 2
        conf = max(0.1, 1.0 - noise / 20)
        if health == NavHealth.FAILED:
            conf = 0.05
        return SensorState(sensor, health, pos_est, vel_est, cov, conf, dt)

    def inject_fault(self, sensor: NavSensor, health: NavHealth):
        self.sensor_health[sensor] = health

    def step(self, dt: float = 0.1) -> NavSolution:
        self.true_pos += self.true_vel * dt + self.rng.standard_normal(3) * 0.01
        self.true_vel += self.rng.standard_normal(3) * 0.05

        self.ekf.predict(dt)
        states = []
        for sensor in [NavSensor.GPS, NavSensor.IMU, NavSensor.VISION]:
            s = self._simulate_sensor(sensor, dt)
            states.append(s)
            if s.health != NavHealth.FAILED:
                H = np.zeros((3, 6))
                H[:3, :3] = np.eye(3)
                R = s.covariance
                self.ekf.update(s.position_est, H, R)

        pos, vel = self.ekf.state()
        voted_pos, integrity = self.voter.vote(states)
        used = [s.sensor for s in states if s.health != NavHealth.FAILED]
        heading = float(np.degrees(np.arctan2(vel[1], vel[0]))) % 360
        confidence = integrity * len(used) / 3

        sol = NavSolution(pos, vel, round(heading, 1), round(confidence, 4),
                         used, round(integrity, 4))
        self.solutions.append(sol)
        return sol

    def run(self, duration: float = 10, dt: float = 0.1) -> List[NavSolution]:
        sols = []
        for _ in np.arange(0, duration, dt):
            sols.append(self.step(dt))
        return sols

    def summary(self) -> Dict:
        return {
            "solutions": len(self.solutions),
            "avg_confidence": round(np.mean([s.confidence for s in self.solutions]), 4) if self.solutions else 0,
            "avg_integrity": round(np.mean([s.integrity for s in self.solutions]), 4) if self.solutions else 0,
            "failed_sensors": sum(1 for h in self.sensor_health.values() if h == NavHealth.FAILED),
        }

File: archive/polyglot/nim/fault_tolerant_nav.nim
# Phase 526: Nim Fault-Tolerant Navigation — EKF Sensor Fusion
import math, strformat

type
  NavSensor = enum
    GPS, IMU, Vision, Barometer, Magnetometer, Lidar

  NavHealth = enum
    Nominal, Degraded, Failed

  Vec3 = object
    x, y, z: float64

  SensorState = object
    sensor: NavSensor
    health: NavHealth
    position: Vec3
    confidence: float64

  NavSolution = object
    position: Vec3
    velocity: Vec3
    heading: float64
    confidence: float64
    integrity: float64
    sensorsUsed: int

  PRNG = object
    state: uint64

proc initPRNG(seed: uint64): PRNG =
  result.state = seed xor 0x6c62272e07bb0142'u64

proc next(rng: var PRNG): uint64 =
  rng.state = rng.state xor (rng.state shl 13)
  rng.state = rng.state xor (rng.state shr 7)
  rng.state = rng.state xor (rng.state shl 17)
  result = rng.state

proc uniform(rng: var PRNG): float64 =
  float64(rng.next() and 0x7FFFFFFF'u64) / float64(0x7FFFFFFF)

proc normal(rng: var PRNG): float64 =
  let u1 = max(rng.uniform(), 1e-10)
  let u2 = rng.uniform()
  sqrt(-2.0 * ln(u1)) * cos(2.0 * PI * u2)

proc vec3(x, y, z: float64): Vec3 = Vec3(x: x, y: y, z: z)
proc add(a, b: Vec3): Vec3 = vec3(a.x+b.x, a.y+b.y, a.z+b.z)
proc scale(a: Vec3, s: float64): Vec3 = vec3(a.x*s, a.y*s, a.z*s)
proc norm(a: Vec3): float64 = sqrt(a.x*a.x + a.y*a.y + a.z*a.z)

type EKF = object
  pos, vel: Vec3
  pDiag: array[6, float64]

proc initEKF(): EKF =
  result.pos = vec3(0, 0, 50)
  result.vel = vec3(2, 1, 0)
  for i in 0..5: result.pDiag[i] = 10.0

proc predict(ekf: var EKF, dt: float64) =
  ekf.pos = add(ekf.pos, scale(ekf.vel, dt))
  for i in 0..5: ekf.pDiag[i] += 0.01 * dt

proc update(ekf: var EKF, measurement: Vec3, noise: float64) =
  let k = ekf.pDiag[0] / (ekf.pDiag[0] + noise * noise)
  ekf.pos.x += k * (measurement.x - ekf.pos.x)
  ekf.pos.y += k * (measurement.y - ekf.pos.y)
  ekf.pos.z += k * (measurement.z - ekf.pos.z)
  for i in 0..2: ekf.pDiag[i] *= (1.0 - k)

proc simulateSensor(rng: var PRNG, truePos: Vec3, sensor: NavSensor,
                     health: NavHealth): SensorState =
  let noiseMap = [2.0, 0.5, 1.0, 3.0, 5.0, 0.3]  # GPS,IMU,Vision,Baro,Mag,Lidar
  var noise = noiseMap[ord(sensor)]
  if health == Degraded: noise *= 5.0
  elif health == Failed: noise *= 50.0

  let pos = vec3(
    truePos.x + rng.normal() * noise,
    truePos.y + rng.normal() * noise,
    truePos.z + rng.normal() * noise
  )
  let conf = max(0.1, 1.0 - noise / 20.0)
  SensorState(sensor: sensor, health: health, position: pos,
              confidence: if health == Failed: 0.05 else: conf)

proc vote(states: seq[SensorState]): (Vec3, float64) =
  var totalW = 0.0
  var pos = vec3(0, 0, 0)
  var healthy = 0
  for s in states:
    if s.health != Failed:
      pos = add(pos, scale(s.position, s.confidence))
      totalW += s.confidence
      healthy += 1
  if totalW > 0:
    pos = scale(pos, 1.0 / totalW)
  let integrity = if healthy >= 2: 0.95 else: 0.5
  (pos, integrity)

when isMainModule:
  var rng = initPRNG(42)
  var ekf = initEKF()
  var truePos = vec3(0, 0, 50)
  var trueVel = vec3(2, 1, 0)
  let sensorHealths = [Nominal, Nominal, Nominal, Nominal, Nominal, Nominal]
  var totalConf = 0.0

  for step in 0..<100:
    let dt = 0.1
    truePos = add(truePos, scale(trueVel, dt))
    trueVel = add(trueVel, vec3(rng.normal()*0.05, rng.normal()*0.05, rng.normal()*0.01))
    ekf.predict(dt)

    var states: seq[SensorState] = @[]
    for sensor in [GPS, IMU, Vision]:
      let health = sensorHealths[ord(sensor)]
      let s = simulateSensor(rng, truePos, sensor, health)
      states.add(s)
      if health != Failed:
        ekf.update(s.position, if health == Degraded: 10.0 else: 2.0)

    let (votedPos, integrity) = vote(states)
    totalConf += integrity

  echo fmt"Nav steps: 100"
  echo fmt"Final pos: ({ekf.pos.x:.1f}, {ekf.pos.y:.1f}, {ekf.pos.z:.1f})"
  echo fmt"True pos:  ({truePos.x:.1f}, {truePos.y:.1f}, {truePos.z:.1f})"
  echo fmt"Avg integrity: {totalConf / 100.0:.4f}"
